community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java

```diff
@@ -679,52 +680,72 @@ public int getDegree()
+        KernelTransaction transaction = safeAcquireTransaction();
+        int typeId = transaction.tokenRead().relationshipType( type.name() );
+        if ( typeId == NO_ID )
+        {   // This type doesn't even exist. Return 0
+            return 0;
         }
-        catch ( EntityNotFoundException e )
+
+        try ( Statement ignore = transaction.acquireStatement() )
         {
-            throw new NotFoundException( "Node not found.", e );
+            NodeCursor nodes = transaction.nodeCursor();
+            singleNode( transaction, nodes );
+            RelationshipGroupCursor group = transaction.cursors().allocateRelationshipGroupCursor();
+
+            return Nodes.countAll( nodes, group, typeId );
         }
     }
 
     @Override
     public int getDegree( Direction direction )
     {
-        try ( Statement statement = spi.statement() )
-        {
-            ReadOperations ops = statement.readOperations();
-            return ops.nodeGetDegree( nodeId, direction );
-        }
-        catch ( EntityNotFoundException e )
+        KernelTransaction transaction = safeAcquireTransaction();
+        try ( Statement ignore = transaction.acquireStatement() )
         {
-            throw new NotFoundException( "Node not found.", e );
+            NodeCursor nodes = transaction.nodeCursor();
+            singleNode( transaction, nodes );
+            RelationshipGroupCursor group = transaction.cursors().allocateRelationshipGroupCursor();
+            switch ( direction )
+            {
+            case OUTGOING:
+                return Nodes.countOutgoing( nodes, group );
+            case INCOMING:
+                return Nodes.countIncoming( nodes, group );
+            case BOTH:
+                return Nodes.countAll( nodes, group );
+            default:
+                throw new IllegalStateException( "Unknown direction " + direction );
+            }
         }
     }
 
     @Override
     public int getDegree( RelationshipType type, Direction direction )
     {
-        try ( Statement statement = spi.statement() )
-        {
-            ReadOperations ops = statement.readOperations();
-            int typeId = ops.relationshipTypeGetForName( type.name() );
-            if ( typeId == NO_ID )
-            {   // This type doesn't even exist. Return 0
-                return 0;
-            }
-            return ops.nodeGetDegree( nodeId, direction, typeId );
+        KernelTransaction transaction = safeAcquireTransaction();
+        int typeId = transaction.tokenRead().relationshipType( type.name() );
+        if ( typeId == NO_ID )
+        {   // This type doesn't even exist. Return 0
+            return 0;
         }
-        catch ( EntityNotFoundException e )
+
+        try ( Statement ignore = transaction.acquireStatement() )
         {
-            throw new NotFoundException( "Node not found.", e );
+            NodeCursor nodes = transaction.nodeCursor();
+            singleNode( transaction, nodes );
```

Comment:
I really like how `singleNode` unifies the `NotFoundException` handling in this class!